### PR TITLE
Fix and refactor

### PR DESCRIPTION
**Pull request by tomasced**
todo
- fix errors following cxl/phpcs-ruleset

changes
- added getDataSourceUUIDbyName()
- fixed bugs (https://app.clickup.com/t/pf4hm5)

### CODE REVIEW

**Comment by lkraav** [comment]
Avoid putting multiple things into single PRs. Always one PR per thing.

File: src/Commands/SubscriptionExportChartMogulCommand.php [hunk 1]
```diff
@@ -45,107 +59,140 @@ class SubscriptionExportChartMogulCommand extends WP_CLI_Command {
...
-     * @throws Exception
+     * @throws \Exception
      */
     public function __construct( array $args, array $assoc_args ) {
 
-        $this->set_command_args( $args, $assoc_args );
+        $ping = $this->initialize_chartmogul();
 
-        $this->initialize_chartmogul();
+        if ( 'pong!' !== $ping ) {
```
Comment (pawelkmpt) [inline src/Commands/SubscriptionExportChartMogulCommand.php:78]: This looks weird. I have a feeling that `ping` should be separate command. But I don't have knowledge about ChartMogul API, so I'm just loud thinking here.

File: src/Commands/SubscriptionExportChartMogulCommand.php [hunk 2]
```diff
@@ -45,107 +59,140 @@ class SubscriptionExportChartMogulCommand extends WP_CLI_Command {
...
+
+            WP_CLI::log( 'No ping to Chartmogul!' );
+            exit;
```
Comment (pawelkmpt) [inline src/Commands/SubscriptionExportChartMogulCommand.php:81]: `WP_CLI::error()` does 2 things:
* displays error message
* exits the script

File: src/Commands/SubscriptionExportChartMogulCommand.php [hunk 3]
```diff
@@ -45,107 +59,140 @@ class SubscriptionExportChartMogulCommand extends WP_CLI_Command {
...
+        }
+
+        $this->set_command_args( $args, $assoc_args );
```
Comment (pawelkmpt) [inline src/Commands/SubscriptionExportChartMogulCommand.php:84]: I haven't seen such approach for WP-CLI. I'd rather check args in command itself.
Also `parse_command_args()` would be a better name

File: src/Commands/SubscriptionExportChartMogulCommand.php [hunk 4]
```diff
@@ -192,143 +252,179 @@ private function create_subscription( int $plan_id, $order_item, $order ) {
...
      */
     private function export_subscriptions(): void {
 
-        if ( ! empty( $this->id ) ) {
+        if ( $this->id > 0 ) {
 
             $subscription = wcs_get_subscription( $this->id );
-            if ( empty( $subscription ) ) {
+
+            if ( ! is_object( $subscription ) ) {
```
Comment (pawelkmpt) [inline src/Commands/SubscriptionExportChartMogulCommand.php:403]: `wcs_get_subscription()` returns `WC_Subscription|false` so better to check:

`if ( ! $subscription )` or `if ( false === $subscription )`

File: src/Commands/SubscriptionExportChartMogulCommand.php [hunk 5]
```diff
@@ -192,143 +252,179 @@ private function create_subscription( int $plan_id, $order_item, $order ) {
...
                 WP_CLI::error( 'Please pass valid subscription id.' );
             }
-
+			
             $this->export_subscription_to_chartmogul( $this->id );
 
             WP_CLI::log( WP_CLI::colorize( '%yExport finished.%n' ) );
 
-        } elseif ( ! empty( $this->fetch_all ) ) {
+        } elseif ( true === $this->fetch_all ) {
```
Comment (pawelkmpt) [inline src/Commands/SubscriptionExportChartMogulCommand.php:411]: just `$this->fetch_all`

File: src/Commands/SubscriptionExportChartMogulCommand.php [hunk 6]
```diff
@@ -192,143 +252,179 @@ private function create_subscription( int $plan_id, $order_item, $order ) {
...
-            if ( empty( $subscription ) ) {
+
+            if ( ! is_object( $subscription ) ) {
                 WP_CLI::error( 'Please pass valid subscription id.' );
             }
-
+			
             $this->export_subscription_to_chartmogul( $this->id );
 
             WP_CLI::log( WP_CLI::colorize( '%yExport finished.%n' ) );
```
Comment (pawelkmpt) [inline src/Commands/SubscriptionExportChartMogulCommand.php:409]: `WP_CLI::success()`

File: src/Commands/SubscriptionExportChartMogulCommand.php [hunk 7]
```diff
@@ -192,143 +252,179 @@ private function create_subscription( int $plan_id, $order_item, $order ) {
...
+
+            WP_CLI::log( 'Total subscriptions: [' . $total_subscriptions . ']' );
+
+            foreach ( $subscriptions as $subscription ) {
 
-            $subscription_ids = $this->get_subscription_posts();
+                $subscription_id = $subscription->get_ID();
+                // $progress_percentage = round( $i / $total_subscriptions * 100 ).
+                // "Progress: $i/$total_subscriptions  [$progress_percentage%]   \r".
+                // ++$i.
```
Comment (pawelkmpt) [inline src/Commands/SubscriptionExportChartMogulCommand.php:424]: There are utils for displaying progress

File: src/Commands/SubscriptionExportChartMogulCommand.php [hunk 8]
```diff
@@ -192,143 +252,179 @@ private function create_subscription( int $plan_id, $order_item, $order ) {
...
+                $this->export_subscription_to_chartmogul( $subscription_id );
 
-            foreach ( $subscription_ids as $post_id ) {
-                // Update post meta.
-                $this->export_subscription_to_chartmogul( $post_id );
+                usleep( 1 );
```
Comment (pawelkmpt) [inline src/Commands/SubscriptionExportChartMogulCommand.php:427]: comment explaining why we sleep here is necessary

File: src/Commands/SubscriptionExportChartMogulCommand.php [hunk 9]
```diff
@@ -192,143 +252,179 @@ private function create_subscription( int $plan_id, $order_item, $order ) {
...
+
+        return ChartMogul\Ping::ping()->data;
     }
 
     /**
      * Export subscriptions to ChartMogul.
-     *
-     * @return void
      */
     private function export_subscriptions(): void {
```
Comment (pawelkmpt) [inline src/Commands/SubscriptionExportChartMogulCommand.php:397]: `export_subscription_to_chartmogul()` takes `int $subscription_id ` as a param. I wonder why in one of if-else blocks we fetch subscription object and in the other one we get post objects instead 🤔

**Comment by lkraav** [comment]
In general, none of the code details matter, until whatever code is there is verified to be able to push subscription data to API and product team is able to identify whether ChartMogul is useful for us in the first place, or not.

So I wouldn't spend any time on code quality here outside of what's already done, unless something literally doesn't work to push subscription data.

Someone needs to take charge of testing whether data goes correctly.

File: src/Commands/SubscriptionExportChartMogulCommand.php [hunk 10]
```diff
@@ -355,43 +451,41 @@ private function get_subscription_posts(): array {
...
+        $customer     = $this->create_customer( $subscription );
+        $orders       = $subscription->get_related_orders();
 
         foreach ( $orders as $order_id ) {
+
             $order = wc_get_order( $order_id );
             $this->create_invoice( $customer, $order );
         }
 
         $this->add_cli_log( $subscription_id );
```
Comment (pawelkmpt) [inline src/Commands/SubscriptionExportChartMogulCommand.php:467]: I'm pretty sure that:

1. `create_invoice()` is too big and too many things happens there
2. `$this->create_invoice()` should be either wrapped with `try-catch` or return something and handled conditionally with `if`
3. it is not obvious what kind of log `$this->add_cli_log()` saves

File: src/Commands/SubscriptionExportChartMogulCommand.php [hunk 11]
```diff
@@ -45,107 +59,140 @@ class SubscriptionExportChartMogulCommand extends WP_CLI_Command {
...
+    /**
+     * Function to set UUID for provided Data Source name.
+     *
+     * @param string $data_source_name ChartMogul Data Source name.
+     * @return string UUID or null value.
+     */
+    private function getDataSourceUUIDbyName( string $data_source_name ): string {
+
+        $all_data_sources   = ChartMogul\DataSource::all();
+        $total_data_sources = count( $all_data_sources );
```
Comment (pawelkmpt) [inline src/Commands/SubscriptionExportChartMogulCommand.php:107]: early return needed

```
if ( 0 === $total_data_sources ) {...}
```

File: src/Commands/SubscriptionExportChartMogulCommand.php [hunk 12]
```diff
@@ -45,107 +59,140 @@ class SubscriptionExportChartMogulCommand extends WP_CLI_Command {
...
+        $i                  = 0;
+
+        for ( ; $i < $total_data_sources; $i++ ) {
```
Comment (pawelkmpt) [inline src/Commands/SubscriptionExportChartMogulCommand.php:110]: it's easier to use `foreach`

File: src/Commands/SubscriptionExportChartMogulCommand.php [hunk 13]
```diff
@@ -45,107 +59,140 @@ class SubscriptionExportChartMogulCommand extends WP_CLI_Command {
...
+            if ( $data_source_name === $all_data_sources[ $i ]->name ) {
+
+                WP_CLI::log( 'Data source ' . $all_data_sources[ $i ]->name . ' UUID: ' . $all_data_sources[ $i ]->uuid );
+                return $all_data_sources[ $i ]->uuid;
+            }
+        }
+
+        return null;
```
Comment (pawelkmpt) [inline src/Commands/SubscriptionExportChartMogulCommand.php:118]: function returns `null` without writing it explicitly

**Comment by kylebrodeur** [comment]
@lkraav I'll be owning if the data is exported to chartmogul is clean. So far, we need more than just subscription data and also the code isn't good.

Here's the plan and notes on issues I have with current code: https://app.clickup.com/2317190/v/dc/26pw6-7581/26pw6-3584

**Review by pawelkmpt — COMMENTED** [review]
I left comments to the most obvious parts. It's not complete code review but it already shows that the code does not meet code quality expectations. There's too many places where script can fail silently, because there's no sanity checks or can fail because of (randomly) missing meta or some other value.

My experience with CLIs which talk to APIs is that you have to check every possible point of failure. Otherwise script can fail in the middle of the processing multiple items and you might not be able to:
* detect why it failed,
* detect where exactly it failed,
* restart from failure point.

General idea of plugin is good. It shows what more-less needs to be done and what SDK methods should be used.

**Comment by pawelkmpt** [comment]
> In general, none of the code details matter, until whatever code is there is verified to be able to push subscription data to API and product team is able to identify whether ChartMogul is useful for us in the first place, or not.

I agree. We should rewrite it only if we decide to use it long term.

**Pull request closed without merging**